dot_local/share/man/man1/rg.1
'\" t
.\"     Title: rg
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.12
.\"      Date: 2021-06-15
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "RG" "1" "2021-06-15" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
rg \- recursively search the current directory for lines matching a pattern
.SH "SYNOPSIS"
.sp
\fBrg\fP [\fIOPTIONS\fP] \fIPATTERN\fP [\fIPATH\fP...]
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-e\fP \fIPATTERN\fP... [\fIPATH\fP...]
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-f\fP \fIPATTERNFILE\fP... [\fIPATH\fP...]
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-\-files\fP [\fIPATH\fP...]
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-\-type\-list\fP
.sp
\fBcommand\fP | \fBrg\fP [\fIOPTIONS\fP] \fIPATTERN\fP
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-\-help\fP
.sp
\fBrg\fP [\fIOPTIONS\fP] \fB\-\-version\fP
.SH "DESCRIPTION"
.sp
ripgrep (rg) recursively searches the current directory for a regex pattern.
By default, ripgrep will respect your .gitignore and automatically skip hidden
files/directories and binary files.
.sp
ripgrep\(cqs default regex engine uses finite automata and guarantees linear
time searching. Because of this, features like backreferences and arbitrary
look\-around are not supported. However, if ripgrep is built with PCRE2, then
the \fB\-\-pcre2\fP flag can be used to enable backreferences and look\-around.
.sp
ripgrep supports configuration files. Set \fBRIPGREP_CONFIG_PATH\fP to a
configuration file. The file can specify one shell argument per line. Lines
starting with \fB#\fP are ignored. For more details, see the man page or the
\fBREADME\fP.
.sp
ripgrep will automatically detect if stdin exists and search stdin for a regex
pattern, e.g. \fBls | rg foo\fP. In some environments, stdin may exist when it
shouldn\(cqt. To turn off stdin detection explicitly specify the directory to
search, e.g. \fBrg foo ./\fP.
.sp
Tip: to disable all smart filtering and make ripgrep behave a bit more like
classical grep, use \fBrg \-uuu\fP.
.SH "REGEX SYNTAX"
.sp
ripgrep uses Rust\(cqs regex engine by default, which documents its syntax:
.URL "https://docs.rs/regex/*/regex/#syntax" "" ""
.sp
ripgrep uses byte\-oriented regexes, which has some additional documentation:
.URL "https://docs.rs/regex/*/regex/bytes/index.html#syntax" "" ""
.sp
To a first approximation, ripgrep uses Perl\-like regexes without look\-around or
backreferences. This makes them very similar to the "extended" (ERE) regular
expressions supported by \fBegrep\fP, but with a few additional features like
Unicode character classes.
.sp
If you\(cqre using ripgrep with the \fB\-\-pcre2\fP flag, then please consult
.URL "https://www.pcre.org" "" " "
or the PCRE2 man pages for documentation on the supported
syntax.
.SH "POSITIONAL ARGUMENTS"
.sp
\fIPATTERN\fP
.RS 4
A regular expression used for searching. To match a pattern beginning with a
dash, use the \-e/\-\-regexp option.
.RE
.sp
\fIPATH\fP
.RS 4
A file or directory to search. Directories are searched recursively. File
paths specified explicitly on the command line override glob and ignore
rules.
.RE
.SH "OPTIONS"
.sp
Note that many options can be disabled via flags. In some cases, those flags
are not listed in a first class way below. For example, the \fB\-\-column\fP
flag (listed below) enables column numbers in ripgrep\(cqs output, but the
\fB\-\-no\-column\fP flag (not listed below) disables them. The reverse can also
exist. For example, the \fB\-\-no\-ignore\fP flag (listed below) disables ripgrep\(cqs
\fBgitignore\fP logic, but the \fB\-\-ignore\fP flag (not listed below) enables it. These
flags are useful for overriding a ripgrep configuration file on the command
line. Each flag\(cqs documentation notes whether an inverted flag exists. In all
cases, the flag specified last takes precedence.
.sp
\fB\-A, \-\-after\-context\fP \fINUM\fP
.RS 4
Show NUM lines after each match.
.sp
This overrides the \-\-context and \-\-passthru flags.
.RE
.sp
\fB\-\-auto\-hybrid\-regex\fP
.RS 4
DEPRECATED. Use \-\-engine instead.
.sp
When this flag is used, ripgrep will dynamically choose between supported regex
engines depending on the features used in a pattern. When ripgrep chooses a
regex engine, it applies that choice for every regex provided to ripgrep (e.g.,
via multiple \-e/\-\-regexp or \-f/\-\-file flags).
.sp
As an example of how this flag might behave, ripgrep will attempt to use
its default finite automata based regex engine whenever the pattern can be
successfully compiled with that regex engine. If PCRE2 is enabled and if the
pattern given could not be compiled with the default regex engine, then PCRE2
will be automatically used for searching. If PCRE2 isn\(cqt available, then this
flag has no effect because there is only one regex engine to choose from.
.sp
In the future, ripgrep may adjust its heuristics for how it decides which
regex engine to use. In general, the heuristics will be limited to a static
analysis of the patterns, and not to any specific runtime behavior observed
while searching files.
.sp
The primary downside of using this flag is that it may not always be obvious
which regex engine ripgrep uses, and thus, the match semantics or performance
profile of ripgrep may subtly and unexpectedly change. However, in many cases,
all regex engines will agree on what constitutes a match and it can be nice
to transparently support more advanced regex features like look\-around and
backreferences without explicitly needing to enable them.
.sp
This flag can be disabled with \-\-no\-auto\-hybrid\-regex.
.RE
.sp
\fB\-B, \-\-before\-context\fP \fINUM\fP
.RS 4
Show NUM lines before each match.
.sp
This overrides the \-\-context and \-\-passthru flags.
.RE
.sp
\fB\-\-binary\fP
.RS 4
Enabling this flag will cause ripgrep to search binary files. By default,
ripgrep attempts to automatically skip binary files in order to improve the
relevance of results and make the search faster.
.sp
Binary files are heuristically detected based on whether they contain a NUL
byte or not. By default (without this flag set), once a NUL byte is seen,
ripgrep will stop searching the file. Usually, NUL bytes occur in the beginning
of most binary files. If a NUL byte occurs after a match, then ripgrep will
still stop searching the rest of the file, but a warning will be printed.
.sp
In contrast, when this flag is provided, ripgrep will continue searching a file
even if a NUL byte is found. In particular, if a NUL byte is found then ripgrep
will continue searching until either a match is found or the end of the file is
reached, whichever comes sooner. If a match is found, then ripgrep will stop
and print a warning saying that the search stopped prematurely.
.sp
If you want ripgrep to search a file without any special NUL byte handling at
all (and potentially print binary data to stdout), then you should use the
\fI\-a/\-\-text\fP flag.
.sp
The \fI\-\-binary\fP flag is a flag for controlling ripgrep\(cqs automatic filtering
mechanism. As such, it does not need to be used when searching a file
explicitly or when searching stdin. That is, it is only applicable when
recursively searching a directory.
.sp
Note that when the \fI\-u/\-\-unrestricted\fP flag is provided for a third time, then
this flag is automatically enabled.
.sp
This flag can be disabled with \fI\-\-no\-binary\fP. It overrides the \fI\-a/\-\-text\fP
flag.
.RE
.sp
\fB\-\-block\-buffered\fP
.RS 4
When enabled, ripgrep will use block buffering. That is, whenever a matching
line is found, it will be written to an in\-memory buffer and will not be
written to stdout until the buffer reaches a certain size. This is the default
when ripgrep\(cqs stdout is redirected to a pipeline or a file. When ripgrep\(cqs
stdout is connected to a terminal, line buffering will be used. Forcing block
buffering can be useful when dumping a large amount of contents to a terminal.
.sp
Forceful block buffering can be disabled with \-\-no\-block\-buffered. Note that
using \-\-no\-block\-buffered causes ripgrep to revert to its default behavior of
automatically detecting the buffering strategy. To force line buffering, use
the \-\-line\-buffered flag.
.RE
.sp
\fB\-b, \-\-byte\-offset\fP
.RS 4
Print the 0\-based byte offset within the input file before each line of output.
If \-o (\-\-only\-matching) is specified, print the offset of the matching part
itself.
.sp
If ripgrep does transcoding, then the byte offset is in terms of the the result
of transcoding and not the original data. This applies similarly to another
transformation on the source, such as decompression or a \-\-pre filter. Note
that when the PCRE2 regex engine is used, then UTF\-8 transcoding is done by
default.
.RE
.sp
\fB\-s, \-\-case\-sensitive\fP
.RS 4
Search case sensitively.
.sp
This overrides the \-i/\-\-ignore\-case and \-S/\-\-smart\-case flags.
.RE
.sp
\fB\-\-color\fP \fIWHEN\fP
.RS 4
This flag controls when to use colors. The default setting is \fIauto\fP, which
means ripgrep will try to guess when to use colors. For example, if ripgrep is
printing to a terminal, then it will use colors, but if it is redirected to a
file or a pipe, then it will suppress color output. ripgrep will suppress color
output in some other circumstances as well. For example, if the TERM
environment variable is not set or set to \fIdumb\fP, then ripgrep will not use
colors.
.sp
The possible values for this flag are:
.sp
.if n .RS 4
.nf
.fam C
never    Colors will never be used.
auto     The default. ripgrep tries to be smart.
always   Colors will always be used regardless of where output is sent.
ansi     Like \(aqalways\(aq, but emits ANSI escapes (even in a Windows console).
.fam
.fi
.if n .RE
.sp
When the \-\-vimgrep flag is given to ripgrep, then the default value for the
\-\-color flag changes to \fInever\fP.
.RE
.sp
\fB\-\-colors\fP \fICOLOR_SPEC\fP ...
.RS 4
This flag specifies color settings for use in the output. This flag may be
provided multiple times. Settings are applied iteratively. Colors are limited
to one of eight choices: red, blue, green, cyan, magenta, yellow, white and
black. Styles are limited to nobold, bold, nointense, intense, nounderline
or underline.
.sp
The format of the flag is \fI{type}:{attribute}:{value}\fP. \fI{type}\fP should be
one of path, line, column or match. \fI{attribute}\fP can be fg, bg or style.
\fI{value}\fP is either a color (for fg and bg) or a text style. A special format,
\fI{type}:none\fP, will clear all color settings for \fI{type}\fP.
.sp
For example, the following command will change the match color to magenta and
the background color for line numbers to yellow:
.sp
.if n .RS 4
.nf
.fam C
rg \-\-colors \(aqmatch:fg:magenta\(aq \-\-colors \(aqline:bg:yellow\(aq foo.
.fam
.fi
.if n .RE
.sp
Extended colors can be used for \fI{value}\fP when the terminal supports ANSI color
sequences. These are specified as either \fIx\fP (256\-color) or \fIx,x,x\fP (24\-bit
truecolor) where x is a number between 0 and 255 inclusive. x may be given as
a normal decimal number or a hexadecimal number, which is prefixed by \f(CR0x\fP.
.sp
For example, the following command will change the match background color to
that represented by the rgb value (0,128,255):
.sp
.if n .RS 4
.nf
.fam C
rg \-\-colors \(aqmatch:bg:0,128,255\(aq
.fam
.fi
.if n .RE
.sp
or, equivalently,
.sp
.if n .RS 4
.nf
.fam C
rg \-\-colors \(aqmatch:bg:0x0,0x80,0xFF\(aq
.fam
.fi
.if n .RE
.sp
Note that the the intense and nointense style flags will have no effect when
used alongside these extended color codes.
.RE
.sp
\fB\-\-column\fP
.RS 4
Show column numbers (1\-based). This only shows the column numbers for the first
match on each line. This does not try to account for Unicode. One byte is equal
to one column. This implies \-\-line\-number.
.sp
This flag can be disabled with \-\-no\-column.
.RE
.sp
\fB\-C, \-\-context\fP \fINUM\fP
.RS 4
Show NUM lines before and after each match. This is equivalent to providing
both the \-B/\-\-before\-context and \-A/\-\-after\-context flags with the same value.
.sp
This overrides both the \-B/\-\-before\-context and \-A/\-\-after\-context flags,
in addition to the \-\-passthru flag.
.RE
.sp
\fB\-\-context\-separator\fP \fISEPARATOR\fP
.RS 4
The string used to separate non\-contiguous context lines in the output. This
is only used when one of the context flags is used (\-A, \-B or \-C). Escape
sequences like \(rsx7F or \(rst may be used. The default value is \-\-.
.sp
When the context separator is set to an empty string, then a line break
is still inserted. To completely disable context separators, use the
\-\-no\-context\-separator flag.
.RE
.sp
\fB\-c, \-\-count\fP
.RS 4
This flag suppresses normal output and shows the number of lines that match
the given patterns for each file searched. Each file containing a match has its
path and count printed on each line. Note that this reports the number of lines
that match and not the total number of matches, unless \-U/\-\-multiline is
enabled. In multiline mode, \-\-count is equivalent to \-\-count\-matches.
.sp
If only one file is given to ripgrep, then only the count is printed if there
is a match. The \-\-with\-filename flag can be used to force printing the file
path in this case. If you need a count to be printed regardless of whether
there is a match, then use \-\-include\-zero.
.sp
This overrides the \-\-count\-matches flag. Note that when \-\-count is combined
with \-\-only\-matching, then ripgrep behaves as if \-\-count\-matches was given.
.RE
.sp
\fB\-\-count\-matches\fP
.RS 4
This flag suppresses normal output and shows the number of individual
matches of the given patterns for each file searched. Each file
containing matches has its path and match count printed on each line.
Note that this reports the total number of individual matches and not
the number of lines that match.
.sp
If only one file is given to ripgrep, then only the count is printed if there
is a match. The \-\-with\-filename flag can be used to force printing the file
path in this case.
.sp
This overrides the \-\-count flag. Note that when \-\-count is combined with
\-\-only\-matching, then ripgrep behaves as if \-\-count\-matches was given.
.RE
.sp
\fB\-\-crlf\fP
.RS 4
When enabled, ripgrep will treat CRLF (\fI\(rsr\(rsn\fP) as a line terminator instead
of just \fI\(rsn\fP.
.sp
Principally, this permits \fI$\fP in regex patterns to match just before CRLF
instead of just before LF. The underlying regex engine may not support this
natively, so ripgrep will translate all instances of \fI$\fP to \fI(?:\(rsr??$)\fP. This
may produce slightly different than desired match offsets. It is intended as a
work\-around until the regex engine supports this natively.
.sp
CRLF support can be disabled with \-\-no\-crlf.
.RE
.sp
\fB\-\-debug\fP
.RS 4
Show debug messages. Please use this when filing a bug report.
.sp
The \-\-debug flag is generally useful for figuring out why ripgrep skipped
searching a particular file. The debug messages should mention all files
skipped and why they were skipped.
.sp
To get even more debug output, use the \-\-trace flag, which implies \-\-debug
along with additional trace data. With \-\-trace, the output could be quite
large and is generally more useful for development.
.RE
.sp
\fB\-\-dfa\-size\-limit\fP \fINUM+SUFFIX?\fP
.RS 4
The upper size limit of the regex DFA. The default limit is 10M. This should
only be changed on very large regex inputs where the (slower) fallback regex
engine may otherwise be used if the limit is reached.
.sp
The argument accepts the same size suffixes as allowed in with the
\-\-max\-filesize flag.
.RE
.sp
\fB\-E, \-\-encoding\fP \fIENCODING\fP
.RS 4
Specify the text encoding that ripgrep will use on all files searched. The
default value is \fIauto\fP, which will cause ripgrep to do a best effort automatic
detection of encoding on a per\-file basis. Automatic detection in this case
only applies to files that begin with a UTF\-8 or UTF\-16 byte\-order mark (BOM).
No other automatic detection is performed. One can also specify \fInone\fP which
will then completely disable BOM sniffing and always result in searching the
raw bytes, including a BOM if it\(cqs present, regardless of its encoding.
.sp
Other supported values can be found in the list of labels here:
.URL "https://encoding.spec.whatwg.org/#concept\-encoding\-get" "" ""
.sp
For more details on encoding and how ripgrep deals with it, see GUIDE.md.
.sp
This flag can be disabled with \-\-no\-encoding.
.RE
.sp
\fB\-\-engine\fP \fIENGINE\fP
.RS 4
Specify which regular expression engine to use. When you choose a regex engine,
it applies that choice for every regex provided to ripgrep (e.g., via multiple
\-e/\-\-regexp or \-f/\-\-file flags).
.sp
Accepted values are \fIdefault\fP, \fIpcre2\fP, or \fIauto\fP.
.sp
The default value is \fIdefault\fP, which is the fastest and should be good for
most use cases. The \fIpcre2\fP engine is generally useful when you want to use
features such as look\-around or backreferences. \fIauto\fP will dynamically choose
between supported regex engines depending on the features used in a pattern on
a best effort basis.
.sp
Note that the \fIpcre2\fP engine is an optional ripgrep feature. If PCRE2 wasn\(cqt
included in your build of ripgrep, then using this flag will result in ripgrep
printing an error message and exiting.
.sp
This overrides previous uses of \-\-pcre2 and \-\-auto\-hybrid\-regex flags.
.RE
.sp
\fB\-\-field\-context\-separator\fP \fISEPARATOR\fP
.RS 4
Set the field context separator, which is used to delimit file paths, line
numbers, columns and the context itself, when printing contextual lines. The
separator may be any number of bytes, including zero. Escape sequences like
\(rsx7F or \(rst may be used. The default value is \-.
.RE
.sp
\fB\-\-field\-match\-separator\fP \fISEPARATOR\fP
.RS 4
Set the field match separator, which is used to delimit file paths, line
numbers, columns and the match itself. The separator may be any number of
bytes, including zero. Escape sequences like \(rsx7F or \(rst may be used. The
default value is \-.
.RE
.sp
\fB\-f, \-\-file\fP \fIPATTERNFILE\fP ...
.RS 4
Search for patterns from the given file, with one pattern per line. When this
flag is used multiple times or in combination with the \-e/\-\-regexp flag,
then all patterns provided are searched. Empty pattern lines will match all
input lines, and the newline is not counted as part of the pattern.
.sp
A line is printed if and only if it matches at least one of the patterns.
.RE
.sp
\fB\-\-files\fP
.RS 4
Print each file that would be searched without actually performing the search.
This is useful to determine whether a particular file is being searched or not.
.RE
.sp
\fB\-l, \-\-files\-with\-matches\fP
.RS 4
Print the paths with at least one match and suppress match contents.
.sp
This overrides \-\-files\-without\-match.
.RE
.sp
\fB\-\-files\-without\-match\fP
.RS 4
Print the paths that contain zero matches and suppress match contents. This
inverts/negates the \-\-files\-with\-matches flag.
.sp
This overrides \-\-files\-with\-matches.
.RE
.sp
\fB\-F, \-\-fixed\-strings\fP
.RS 4
Treat the pattern as a literal string instead of a regular expression. When
this flag is used, special regular expression meta characters such as .(){}*+
do not need to be escaped.
.sp
This flag can be disabled with \-\-no\-fixed\-strings.
.RE
.sp
\fB\-L, \-\-follow\fP
.RS 4
When this flag is enabled, ripgrep will follow symbolic links while traversing
directories. This is disabled by default. Note that ripgrep will check for
symbolic link loops and report errors if it finds one.
.sp
This flag can be disabled with \-\-no\-follow.
.RE
.sp
\fB\-g, \-\-glob\fP \fIGLOB\fP ...
.RS 4
Include or exclude files and directories for searching that match the given
glob. This always overrides any other ignore logic. Multiple glob flags may be
used. Globbing rules match .gitignore globs. Precede a glob with a ! to exclude
it. If multiple globs match a file or directory, the glob given later in the
command line takes precedence.
.sp
As an extension, globs support specifying alternatives: \fB\-g ab{c,d}\fP is
equivalet to \fB\-g abc \-g abd\fP. Empty alternatives like \fB\-g ab{,c}\fP are not
currently supported. Note that this syntax extension is also currently enabled
in gitignore files, even though this syntax isn\(cqt supported by git itself.
ripgrep may disable this syntax extension in gitignore files, but it will
always remain available via the \-g/\-\-glob flag.
.sp
When this flag is set, every file and directory is applied to it to test for
a match. So for example, if you only want to search in a particular directory
\fIfoo\fP, then \fB\-g foo\fP is incorrect because \fIfoo/bar\fP does not match the glob
\fIfoo\fP. Instead, you should use \fB\-g \(aqfoo/**\(aq\fP.
.RE
.sp
\fB\-\-glob\-case\-insensitive\fP
.RS 4
Process glob patterns given with the \-g/\-\-glob flag case insensitively. This
effectively treats \-\-glob as \-\-iglob.
.sp
This flag can be disabled with the \-\-no\-glob\-case\-insensitive flag.
.RE
.sp
\fB\-\-heading\fP
.RS 4
This flag prints the file path above clusters of matches from each file instead
of printing the file path as a prefix for each matched line. This is the
default mode when printing to a terminal.
.sp
This overrides the \-\-no\-heading flag.
.RE
.sp
\fB\-., \-\-hidden\fP
.RS 4
Search hidden files and directories. By default, hidden files and directories
are skipped. Note that if a hidden file or a directory is whitelisted in an
ignore file, then it will be searched even if this flag isn\(cqt provided.
.sp
A file or directory is considered hidden if its base name starts with a dot
character (\fI.\fP). On operating systems which support a \f(CRhidden\fP file attribute,
like Windows, files with this attribute are also considered hidden.
.sp
This flag can be disabled with \-\-no\-hidden.
.RE
.sp
\fB\-\-iglob\fP \fIGLOB\fP ...
.RS 4
Include or exclude files and directories for searching that match the given
glob. This always overrides any other ignore logic. Multiple glob flags may be
used. Globbing rules match .gitignore globs. Precede a glob with a ! to exclude
it. Globs are matched case insensitively.
.RE
.sp
\fB\-i, \-\-ignore\-case\fP
.RS 4
When this flag is provided, the given patterns will be searched case
insensitively. The case insensitivity rules used by ripgrep conform to
Unicode\(cqs "simple" case folding rules.
.sp
This flag overrides \-s/\-\-case\-sensitive and \-S/\-\-smart\-case.
.RE
.sp
\fB\-\-ignore\-file\fP \fIPATH\fP ...
.RS 4
Specifies a path to one or more .gitignore format rules files. These patterns
are applied after the patterns found in .gitignore and .ignore are applied
and are matched relative to the current working directory. Multiple additional
ignore files can be specified by using the \-\-ignore\-file flag several times.
When specifying multiple ignore files, earlier files have lower precedence
than later files.
.sp
If you are looking for a way to include or exclude files and directories
directly on the command line, then used \-g instead.
.RE
.sp
\fB\-\-ignore\-file\-case\-insensitive\fP
.RS 4
Process ignore files (.gitignore, .ignore, etc.) case insensitively. Note that
this comes with a performance penalty and is most useful on case insensitive
file systems (such as Windows).
.sp
This flag can be disabled with the \-\-no\-ignore\-file\-case\-insensitive flag.
.RE
.sp
\fB\-\-include\-zero\fP
.RS 4
When used with \-\-count or \-\-count\-matches, print the number of matches for
each file even if there were zero matches. This is disabled by default but can
be enabled to make ripgrep behave more like grep.
.RE
.sp
\fB\-v, \-\-invert\-match\fP
.RS 4
Invert matching. Show lines that do not match the given patterns.
.RE
.sp
\fB\-\-json\fP
.RS 4
Enable printing results in a JSON Lines format.
.sp
When this flag is provided, ripgrep will emit a sequence of messages, each
encoded as a JSON object, where there are five different message types:
.sp
\fBbegin\fP \- A message that indicates a file is being searched and contains at
least one match.
.sp
\fBend\fP \- A message the indicates a file is done being searched. This message
also include summary statistics about the search for a particular file.
.sp
\fBmatch\fP \- A message that indicates a match was found. This includes the text
and offsets of the match.
.sp
\fBcontext\fP \- A message that indicates a contextual line was found. This
includes the text of the line, along with any match information if the search
was inverted.
.sp
\fBsummary\fP \- The final message emitted by ripgrep that contains summary
statistics about the search across all files.
.sp
Since file paths or the contents of files are not guaranteed to be valid UTF\-8
and JSON itself must be representable by a Unicode encoding, ripgrep will emit
all data elements as objects with one of two keys: \fItext\fP or \fIbytes\fP. \fItext\fP is
a normal JSON string when the data is valid UTF\-8 while \fIbytes\fP is the base64
encoded contents of the data.
.sp
The JSON Lines format is only supported for showing search results. It cannot
be used with other flags that emit other types of output, such as \-\-files,
\-\-files\-with\-matches, \-\-files\-without\-match, \-\-count or \-\-count\-matches.
ripgrep will report an error if any of the aforementioned flags are used in
concert with \-\-json.
.sp
Other flags that control aspects of the standard output such as
\-\-only\-matching, \-\-heading, \-\-replace, \-\-max\-columns, etc., have no effect
when \-\-json is set.
.sp
A more complete description of the JSON format used can be found here:
.URL "https://docs.rs/grep\-printer/*/grep_printer/struct.JSON.html" "" ""
.sp
The JSON Lines format can be disabled with \-\-no\-json.
.RE
.sp
\fB\-\-line\-buffered\fP
.RS 4
When enabled, ripgrep will use line buffering. That is, whenever a matching
line is found, it will be flushed to stdout immediately. This is the default
when ripgrep\(cqs stdout is connected to a terminal, but otherwise, ripgrep will
use block buffering, which is typically faster. This flag forces ripgrep to
use line buffering even if it would otherwise use block buffering. This is
typically useful in shell pipelines, e.g.,
\fItail \-f something.log | rg foo \-\-line\-buffered | rg bar\fP.
.sp
Forceful line buffering can be disabled with \-\-no\-line\-buffered. Note that
using \-\-no\-line\-buffered causes ripgrep to revert to its default behavior of
automatically detecting the buffering strategy. To force block buffering, use
the \-\-block\-buffered flag.
.RE
.sp
\fB\-n, \-\-line\-number\fP
.RS 4

dot_local/share/bash-completion/completions/rg.bash
_rg() {
    local i cur prev opts cmds
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    cmd=""
    opts=""

    for i in ${COMP_WORDS[@]}
    do
        case "${i}" in
            rg)
                cmd="rg"
                ;;
            
            *)
                ;;
        esac
    done

    case "${cmd}" in
        rg)
            opts=" -b -s -c -l -F -L -. -i -v -n -N -x -U -0 -o -P -p -q -z -S -a -u -H -I -w -h -V -A -B -C -E -f -g -M -m -e -r -j -t -T  --auto-hybrid-regex --no-auto-hybrid-regex --binary --no-binary --block-buffered --no-block-buffered --byte-offset --case-sensitive --column --no-column --no-context-separator --count --count-matches --crlf --no-crlf --debug --trace --no-encoding --files --files-with-matches --files-without-match --fixed-strings --no-fixed-strings --follow --no-follow --glob-case-insensitive --no-glob-case-insensitive --heading --no-heading --hidden --no-hidden --ignore-case --ignore-file-case-insensitive --no-ignore-file-case-insensitive --include-zero --invert-match --json --no-json --line-buffered --no-line-buffered --line-number --no-line-number --line-regexp --max-columns-preview --no-max-columns-preview --mmap --no-mmap --multiline --no-multiline --multiline-dotall --no-multiline-dotall --no-config --no-ignore --ignore --no-ignore-dot --ignore-dot --no-ignore-exclude --ignore-exclude --no-ignore-files --ignore-files --no-ignore-global --ignore-global --no-ignore-messages --ignore-messages --no-ignore-parent --ignore-parent --no-ignore-vcs --ignore-vcs --no-messages --messages --no-pcre2-unicode --pcre2-unicode --no-require-git --require-git --no-unicode --unicode --null --null-data --one-file-system --no-one-file-system --only-matching --passthru --pcre2 --no-pcre2 --pcre2-version --no-pre --pretty --quiet --search-zip --no-search-zip --smart-case --sort-files --no-sort-files --stats --no-stats --text --no-text --trim --no-trim --type-list --unrestricted --vimgrep --with-filename --no-filename --word-regexp --help --version --after-context --before-context --color --colors --context --context-separator --dfa-size-limit --encoding --engine --field-context-separator --field-match-separator --file --glob --iglob --ignore-file --max-columns --max-count --max-depth --max-filesize --path-separator --pre --pre-glob --regex-size-limit --regexp --replace --sort --sortr --threads --type --type-add --type-clear --type-not  <PATTERN> <PATH>... "
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                
                --after-context)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -A)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --before-context)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -B)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --color)
                    COMPREPLY=($(compgen -W "never auto always ansi" -- "${cur}"))
                    return 0
                    ;;
                --colors)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --context)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -C)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --context-separator)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --dfa-size-limit)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --encoding)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -E)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --engine)
                    COMPREPLY=($(compgen -W "default pcre2 auto" -- "${cur}"))
                    return 0
                    ;;
                --field-context-separator)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --field-match-separator)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --file)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -f)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --glob)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -g)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --iglob)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --ignore-file)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --max-columns)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -M)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --max-count)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -m)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --max-depth)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --max-filesize)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --path-separator)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --pre)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --pre-glob)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --regex-size-limit)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --regexp)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -e)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --replace)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -r)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --sort)
                    COMPREPLY=($(compgen -W "path modified accessed created none" -- "${cur}"))
                    return 0
                    ;;
                --sortr)
                    COMPREPLY=($(compgen -W "path modified accessed created none" -- "${cur}"))
                    return 0
                    ;;
                --threads)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -j)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --type)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -t)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --type-add)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --type-clear)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --type-not)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                    -T)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        
    esac
}

complete -F _rg -o bashdefault -o default rg
